### PR TITLE
Refactored pool heuristic to return 0 or minimum pool size

### PR DESCRIPTION
**Pull request by mcfadden8**

### CODE REVIEW

File: src/umpire/strategy/DynamicPoolMap.cpp [hunk 1]
```diff
@@ -253,25 +255,26 @@ void DynamicPoolMap::coalesce()
...
+    std::size_t size_post{getActualSize()};
 
     // If this removed anything from the map, re-allocate a single large chunk
     // and insert to free map
-    if (released_bytes > 0) {
-      UMPIRE_LOG(Debug, "coalescing " << released_bytes << " bytes.");
-      Pointer ptr{allocateBlock(released_bytes)};
-      insertFree(ptr, released_bytes, true, released_bytes);
+    if (size_post < suggested_size) {
+      std::size_t alloc_size{suggested_size - size_post};
```
Comment (davidbeckingsale) [inline src/umpire/strategy/DynamicPoolMap.cpp:274]: Do you think we should check if alloc_size is reasonable?

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:274]: Good question. What is reasonable?

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPoolMap.cpp:274]: I think it should be _at least_ the minimum block size (I think that will happen anyway, but good to check)

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:274]: @davidbeckingsale, I am assuming we resolved this in our discussion yesterday.  Please let me know if you are still looking for changes.  Thx

Comment (davidbeckingsale) [reply src/umpire/strategy/DynamicPoolMap.cpp:274]: I think the decision was made. I don't think we need to enforce anything, but can you please add a log message here that prints the alloc_size to Debug?

Comment (mcfadden8) [reply src/umpire/strategy/DynamicPoolMap.cpp:274]: Do you mean like on line 275?

**Review by davidbeckingsale — APPROVED** [review]

File: tools/replay/ReplayOperationManager.cpp [hunk 2]
```diff
@@ -14,7 +14,9 @@
 #include "umpire/Allocator.hpp"
 #include "umpire/strategy/AllocationAdvisor.hpp"
 #include "umpire/strategy/AllocationPrefetcher.hpp"
+#include "umpire/strategy/PoolCoalesceHeuristic.hpp"
 #include "umpire/strategy/SizeLimiter.hpp"
+#include "umpire/strategy/QuickPool.hpp"
```
Comment (kab163) [inline tools/replay/ReplayOperationManager.cpp:19]: Seeing as you didn't also need to include DynamicPoolList here, is adding QuickPool necessary?

File: src/umpire/strategy/DynamicSizePool.hpp [hunk 3]
```diff
@@ -432,13 +432,17 @@ class DynamicSizePool : private umpire::strategy::mixins::AlignedAllocation {
     return nb;
   }
 
-  void coalesce()
+  void coalesce(std::size_t suggested_size)
   {
     if (getFreeBlocks() > 1) {
-      std::size_t size_to_coalesce = freeReleasedBlocks();
+      freeReleasedBlocks();
```
Comment (kab163) [inline src/umpire/strategy/DynamicSizePool.hpp:438]: Maybe I'm missing something, but I'm confused how before this function returned some size_t value and now there is no return value (or it is being ignored). Do you see what I mean?

**Review by kab163 — COMMENTED** [review]
Looks good, just had a few questions. This is exciting though :)

File: src/umpire/strategy/DynamicSizePool.hpp (hunk 3 above)
Comment (mcfadden8) [reply src/umpire/strategy/DynamicSizePool.hpp:438]: Good question.  `freeReleasedBlocks()` returns the number of bytes that were freed.  The coalesce method is now concerned about what the total size of the pool should be.  So, we call `getActualSize()` after `freeReleasedBlocks()` to determine the new size of the pool.  If that size is smaller than `suggested_size`, we then allocate a block big enough to increase the pools size.

Comment (kab163) [reply src/umpire/strategy/DynamicSizePool.hpp:438]: Hmm I see. So it is purposely ignored.. could the number of bytes that were freed be useful at all (like in the logging statement that follows)? But ya, that makes sense. Thanks for clarifying!

**Pull request merged**